### PR TITLE
[jsk_2016_01_baxter_apc] work_order modified to apply for apc2016 objects

### PR DESCRIPTION
**Pull request by knorth55**
MERGE after #1546

### CODE REVIEW

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py [hunk 1]
```diff
@@ -5,24 +5,29 @@
...
-    for bin_, n_contents in sorted(bin_n_contents.items(), key=lambda x: x[1]):
-        if n_contents > 5:  # Level3
-            continue
-        sorted_work_order.append((bin_, work_order[bin_]))
+    object_data = get_object_data()
+    sorted_bin_list = bin_contents.keys()
+    if apc2016:
+        sorted_bin_list = sorted(sorted_bin_list,
+                                 key=lambda bin_: filter(lambda x: x['name'] == work_order[bin_],
+                                                         object_data)[0]['graspability'][0][gripper])
```
Comment (yuyu2172) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: Unreadable.

`filter(...)` can be rewritten using Python's list comprehension.

```
filter([x['name'] == work_order[bin_] for x in object_data])
```

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py [hunk 2]
```diff
@@ -36,14 +41,26 @@ def get_work_order_msg(json_file):
...
     for bin_, target_object in work_order:
-        if target_object in abandon_target_objects:
-            continue
-        bin_contents_bool = False
-        for bin_object in bin_contents[bin_]:
-            if bin_object in abandon_bin_objects:
-                bin_contents_bool = True
-        if bin_contents_bool:
+        if apc2016:
+            target_object_data = filter(lambda x: x['name'] == target_object, object_data)
```
Comment (yuyu2172) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]:
```
([x for x in object_data if x['name'] == target_object])
```

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py [hunk 3]
```diff
@@ -5,24 +5,29 @@
...
+                                 key=lambda bin_: filter(lambda x: x['name'] == work_order[bin_],
+                                                         object_data)[0]['graspability'][0][gripper])
+    sorted_bin_list = sorted(sorted_bin_list, key=lambda bin_: len(bin_contents[bin_]))
+    sorted_work_order = [(bin_, work_order[bin_]) for bin_ in sorted_bin_list]
     return sorted_work_order
 
 
-def get_work_order_msg(json_file):
-    work_order = get_sorted_work_order(json_file=json_file)
+def get_work_order_msg(json_file, apc2016):
```
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: `apc2016` flag is necessary? You can check the existence of `'graspability'`, I think.

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 1 above)
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: This cannot be like below?
(I haven't check actually, so please revise if you have some errors)
```python
def get_bin_graspability(bin_name):
    target_obj = work_order[bin_name]
    graspability = object_data[target_obj]['graspability'][gripper]
    return graspability

sorted_bin_list = sorted(bin_contents.keys(), key=get_graspability)
```

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 2 above)
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: i dont like that style

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 3 above)
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: we need it because we have to distinguish `apc2016` with json file, and json style is same both in `apc2015` and `apc2016`.

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 2 above)
Comment (yuyu2172) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: I thought you would say that. lol

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 1 above)
Comment (yuyu2172) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: *much* better

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 2 above)
Comment (yuyu2172) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: some relevant discussions.

TLDR
http://stackoverflow.com/questions/3013449/list-filtering-list-comprehension-vs-lambda-filter

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 3 above)
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: > we have to distinguish apc2016 with json file,

why>

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 2 above)
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: I like that style

> target_object_data = [x for x in object_data if x['name'] == target_object]

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 3 above)
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: because of graspability and so on.

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py [hunk 4]
```diff
@@ -36,14 +41,26 @@ def get_work_order_msg(json_file):
...
+                    continue
+            else:
+                continue
+        else:
+            if target_object in abandon_target_objects:
+                continue
+            bin_contents_bool = False
+            for bin_object in bin_contents[bin_]:
+                if bin_object in abandon_bin_objects:
+                    bin_contents_bool = True
```
Comment (yuyu2172) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]:
```
bin_contents_bool = any([bin_object in abandon_bin_objects for bin_object in bin_contents[bin_]])
```

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 3 above)
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: Why below does not work?

>  You can check the existence of 'graspability' key in dict, I think.

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py [hunk 5]
```diff
@@ -36,14 +41,26 @@ def get_work_order_msg(json_file):
...
+            if bin_contents_bool:
+                continue
```
Comment (yuyu2172) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: from line 57 to 62.
You can delete `bin_contents_bool`, which is unnecessary.
```
if any([bin_object in abandon_bin_objects for bin_object in bin_contents[bin_]]):
    continue
```

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 2 above)
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: object_data is dict, not list

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 3 above)
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: it check object_data.yaml

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 2 above)
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: that can be below: but it's ok.
```
target_obj_data = [entry for entry in obj_data if entry['name'] == target_obj][0]
```

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 3 above)
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: ?

Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: but you don't use `jsk_apc2016_common.get_object_data_2015`, do you??

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 2 above)
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: i see

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py [hunk 6]
```diff
@@ -5,24 +5,35 @@
...
-    for bin_, n_contents in sorted(bin_n_contents.items(), key=lambda x: x[1]):
-        if n_contents > 5:  # Level3
-            continue
-        sorted_work_order.append((bin_, work_order[bin_]))
+    object_data = get_object_data()
+    sorted_bin_list = bin_contents.keys()
+
+    def get_graspability(bin_):
+        target_object = work_order[bin_]
+        target_object_data = filter(lambda object_data: object_data['name'] == target_object, object_data)
```
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: Please add `[0}` at the end.

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py [hunk 7]
```diff
@@ -5,24 +5,35 @@
...
+        graspability = target_object_data[0]['graspability'][0][gripper]
```
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: the second `[0]` is redundant. Please fix the yaml file.

Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: ok

File: jsk_2016_01_baxter_apc/node_scripts/work_order.py (hunk 3 above)
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/work_order.py]: i mean object_data_2016.yaml, but it is not mainstreamof this discussion

**Comment by knorth55** [comment]
@yuyu2172 check it

**Comment by wkentaro** [comment]
@knorth55 Please refer to #1547 
That contains what I wanted you to change.

**Comment by knorth55** [comment]
@wkentaro what is it for?

**Comment by knorth55** [comment]
please tell me what you want me to do in words.

File: jsk_2016_01_baxter_apc/launch/main.launch [hunk 8]
```diff
@@ -9,7 +9,10 @@
 
   <!-- work order data -->
   <node pkg="jsk_2016_01_baxter_apc" type="work_order.py" name="work_order" respawn="true">
-    <param name="json" value="$(arg json)" />
+    <rosparam>
+        json: $(arg json) 
+        is_apc2016: true
```
Comment (wkentaro) [inline jsk_2016_01_baxter_apc/launch/main.launch]: is_apc2016: false at this time? untile ver0.8 is over.

Comment (knorth55) [inline jsk_2016_01_baxter_apc/launch/main.launch]: what is ver 0.8? i have never heard of it.
is it for demo tomorrow?

Comment (wkentaro) [inline jsk_2016_01_baxter_apc/launch/main.launch]: Sorry about that. I mean demo for tomorrow and the day after tomorrow.

Comment (knorth55) [inline jsk_2016_01_baxter_apc/launch/main.launch]: ok, i see

Comment (knorth55) [inline jsk_2016_01_baxter_apc/launch/main.launch]: #1524
